app/bundles/LeadBundle/EventListener/ReportNormalizeSubscriber.php

```diff
@@ -0,0 +1,65 @@
+<?php
+
+namespace Mautic\LeadBundle\EventListener;
+
+use Mautic\LeadBundle\Helper\CustomFieldValueHelper;
+use Mautic\LeadBundle\Model\CompanyModel;
+use Mautic\LeadBundle\Model\LeadModel;
+use Mautic\ReportBundle\Event\ReportDataEvent;
+use Mautic\ReportBundle\ReportEvents;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+
+class ReportNormalizeSubscriber implements EventSubscriberInterface
+{
+    private CompanyModel $companyModel;
+
+    private LeadModel $leadModel;
+
+    public function __construct(LeadModel $leadModel, CompanyModel $companyModel)
+    {
+        $this->leadModel    = $leadModel;
+        $this->companyModel = $companyModel;
+    }
+
+    /**
+     * @return array<string, array<int, int|string>>
+     */
+    public static function getSubscribedEvents(): array
+    {
+        return [
+            ReportEvents::REPORT_ON_DISPLAY => ['onReportDisplay', 0],
+        ];
+    }
+
+    public function onReportDisplay(ReportDataEvent $event): void
+    {
+        if (!$this->useContactOrCompanyColumn($event->getReport()->getColumns())) {
+            return;
+        }
+
+        $fields = array_merge($this->leadModel->getRepository()->getCustomFieldList('lead')[0], $this->companyModel->getRepository()->getCustomFieldList('company')[0]);
```

Comment:
Can this be replaced with 1 specialized query (instead of the 2) that will fetch all custom fields for both objects and only the field types that need normalization? Reports are slow as is. Every new query will make the pain only greater.